### PR TITLE
Fix sftp delete implementation

### PR DESCRIPTION
**Pull request by sevein**
Quick follow-up for https://github.com/artefactual-sdps/enduro/pull/772.

### CODE REVIEW

File: internal/sftp/goclient.go [hunk 1]
```diff
@@ -64,23 +64,29 @@ func (c *GoClient) Upload(ctx context.Context, src io.Reader, dest string) (int6
...
-			return fmt.Errorf("SFTP: insufficient permissions to delete file: %w", err)
+	// SFTP assumes that "/" is used as the directory separator. See:
+	// https://datatracker.ietf.org/doc/html/draft-ietf-secsh-filexfer-02#section-6.2
+	remotePath := strings.TrimSuffix(c.cfg.RemoteDir, "/") + "/" + dest
+
+	if err := c.sftp.Remove(remotePath); err != nil {
+		head := fmt.Sprintf("SFTP: unable to remove file %q", dest)
+		if os.IsNotExist(err) || os.IsPermission(err) {
+			return fmt.Errorf("%s: %w", head, err)
+		}
```
Comment (sevein) [inline internal/sftp/goclient.go]:
```
SFTP: unable to remove file "test.txt": permission denied
SFTP: unable to remove file "test.txt": file does not exist
```

File: internal/sftp/goclient.go [hunk 2]
```diff
@@ -64,23 +64,29 @@ func (c *GoClient) Upload(ctx context.Context, src io.Reader, dest string) (int6
...
+		if status, ok := err.(*sftp.StatusError); ok {
+			return fmt.Errorf("%s: %v (status code: %d)", head, status.FxCode(), status.Code)
```
Comment (sevein) [inline internal/sftp/goclient.go]:
```
SFTP: unable to remove file "restricted/test.txt": failure (status code: 4)
```

File: internal/sftp/goclient.go [hunk 3]
```diff
@@ -64,23 +64,29 @@ func (c *GoClient) Upload(ctx context.Context, src io.Reader, dest string) (int6
...
 		}
-		return fmt.Errorf("SFTP: unable to remove file %q: %v", dest, err)
+		return fmt.Errorf("%s: %v", dest, err)
```
Comment (sevein) [inline internal/sftp/goclient.go]: This shouldn't happen:

    SFTP: unable to remove file "test.txt": ...

File: internal/sftp/goclient_test.go [hunk 4]
```diff
@@ -314,78 +314,93 @@ func UploadTest(t *testing.T) {
...
 	if err != nil {
 		t.Fatalf("Bad server address: %s", serverAddress)
 	}
 
 	_ = startSFTPServer(t, serverAddress)
 
+	type params struct {
+		fsOps       []tfs.PathOp // The state of the filesystem served by the SFTP server.
+		restrictDir string       // Set 0o555 on dir to reproduce permission issues.
+		delete      string       // The file that we will delete.
```
Comment (Diogenesoftoronto) [inline internal/sftp/goclient_test.go]: we could just call it file

File: internal/sftp/goclient.go [hunk 5]
```diff
@@ -64,23 +64,29 @@ func (c *GoClient) Upload(ctx context.Context, src io.Reader, dest string) (int6
...
 	}
+
```
Comment (Diogenesoftoronto) [inline internal/sftp/goclient.go:92]: Why add this change?

File: internal/sftp/goclient.go [hunk 6]
```diff
@@ -64,23 +64,29 @@ func (c *GoClient) Upload(ctx context.Context, src io.Reader, dest string) (int6
...
 
-	// Delete the file
-	if err := c.sftp.Remove(dest); err != nil {
-		if os.IsNotExist(err) {
-			return fmt.Errorf("SFTP: file does not exist: %w", err)
-		} else if os.IsPermission(err) {
-			return fmt.Errorf("SFTP: insufficient permissions to delete file: %w", err)
+	// SFTP assumes that "/" is used as the directory separator. See:
+	// https://datatracker.ietf.org/doc/html/draft-ietf-secsh-filexfer-02#section-6.2
+	remotePath := strings.TrimSuffix(c.cfg.RemoteDir, "/") + "/" + dest
```
Comment (Diogenesoftoronto) [inline internal/sftp/goclient.go:80]: Should we not just use os path separator?

**Review by Diogenesoftoronto — CHANGES_REQUESTED** [review]

File: internal/sftp/goclient.go (hunk 5 above)
Comment (sevein) [reply internal/sftp/goclient.go:92]: For consistency with the other final returns in this file.

File: internal/sftp/goclient.go (hunk 6 above)
Comment (sevein) [reply internal/sftp/goclient.go:80]: The comment suggests that it's a slash only, i.e. not a back-slash.

File: internal/sftp/goclient.go (hunk 3 above)
Comment (djjuhasz) [reply internal/sftp/goclient.go]: It *shouldn't* happen, but if it did happen I think `return fmt.Errorf("SFTP: remove: %s: %v", dest, err)` (or something similar) would provide helpful context for tracking down where the error came from in the code.

File: internal/sftp/goclient.go (hunk 2 above)
Comment (djjuhasz) [reply internal/sftp/goclient.go]: The "failure (status code: 4)" message here isn't very helpful - maybe `status.Error()` would have more information about the cause of the error?

**Review by djjuhasz — APPROVED** [review]
Looks good to me @sevein. I added a couple of comments about the error reporting, but they aren't blockers.

File: internal/sftp/goclient.go (hunk 3 above)
Comment (sevein) [reply internal/sftp/goclient.go]: Yes, good point, fixed in d89087b2.

File: internal/sftp/goclient_test.go (hunk 4 above)
Comment (sevein) [reply internal/sftp/goclient_test.go]: Thanks, fixed in 4816cad7.

File: internal/sftp/goclient.go (hunk 2 above)
Comment (sevein) [reply internal/sftp/goclient.go]: Tried to improve it in 0a613cb1, let me know.

Comment (djjuhasz) [reply internal/sftp/goclient.go]: @sevein I think there is something weird going on with the error handling in the sftp package. :(

The full `Error()` message returned from your "Errors when there are insufficient permissions" test is:

```
"remove /tmp/sftp_test_remote-767072262/restricted/test.txt: permission denied" (SSH_FX_FAILURE)
```

But I would expect a permission denied error to be caught by the `os.IsPermission(err)` check and not make it to this line. :confused: I think something weird is happening at https://github.com/pkg/sftp/blob/v1.13.6/client.go#L741 but their tests don't cover the `RemoveDirectory()` method at all. :(

In any case, I think "failure", "status code: 4", and "SSH_FX_FAILURE" are all generic errors that are not very helpful, but maybe it's the best we can do right now.  I don't really have a preference between "(status code: 4)" or "(SSH_FX_FAILURE)".

Comment (djjuhasz) [reply internal/sftp/goclient.go]: Looks like error handling in `sftp.Remove()` is a thorny issue: https://github.com/pkg/sftp/issues/468

Comment (djjuhasz) [reply internal/sftp/goclient.go]: Side note: the help text for `os.IsPermission(err)` recommends using `errors.Is(err, fs.ErrPermission)` instead:

```
This function predates errors.Is. It only supports errors returned by the os package. New code should use errors.Is(err, fs.ErrPermission).
```

`os.IsNotExist(err)` likewise recommends `errors.Is(err, fs.ErrNotExist)`.

**Review by Diogenesoftoronto — APPROVED** [review]

File: internal/sftp/goclient.go (hunk 2 above)
Comment (sevein) [reply internal/sftp/goclient.go]: Yeah I noticed that too, that's why I ended up working on looking up the StatusError, so weird.

Ok, I'll use errors.Is, good catch!

**Pull request merged**